README.md
ng-bubble-cloud
===============

Simple [bubble-cloud][bubble cloud] directive for [AngularJS][] + [D3][]. 

Get started:

```js
$scope.net_worth = {
    assets: [
        { name: 'Checking & Savings', balance: 2000 },
        { name: 'Money Market', balance: 11000 },
        { name: 'Retirement account', balance: 28000 },
        { name: 'Car', balance: 3000 },
    ],
	liabilities: [
        { name: 'Credit Cards', balance: 5000 },
        { name: 'Laptop', balance: 1500 },
        { name: 'Student Loans', balance: 35000 },
    ],
};

$scope.fill_color = function (group) {
    return group === 'assets' ? '#3eb536' : '#641726';
}

$scope.label_color = function (group) {
    return group === 'assets' ? 'black' : '#8c8c8c';
}
```

```html
<bubble-cloud
	diameter="480"
	watch="true"
	data="net_worth"
	value-attr="balance"
	label-attr="name"
	fill-color-fn="fill_color"
	label-color-fn="label_color"
></bubble>
```

![Net worth example][]

Features
--------

- Wraps up a bubble cloud layout
- Bind to data in your controller
- Chart updates as data changes
- Customize formatting with controller functions
- Chart data can be grouped hierarchically or using attributes


Installation
------------

Install ng-bubble-cloud by running:

	bower install ng-bubble-cloud


```html
<script src="path/to/bower_components/underscore/underscore.js"></script>
<script src="path/to/bower_components/angular/angular.js"></script>
<script src="path/to/bower_components/d3/d3.js"></script>
<script src="path/to/bower_components/ng-bubble-cloud/src/bubble-cloud.js"></script>
```


Contribute
----------

- [Issue tracker][issues]
- [Source code][source]


Support
-------

If you are having problems, please open an issue.


License
-------

The project is licensed under the MIT license.


[bubble cloud]: http://bl.ocks.org/mbostock/4063269
[AngularJS]: https://angularjs.org/
[D3]: http://d3js.org/
[Net worth example]: http://paulmelnikow.github.io/ng-bubble-cloud/net_worth.png
[issues]: https://github.com/paulmelnikow/ng-bubble-cloud/issues
[source]: https://github.com/paulmelnikow/ng-bubble-cloud


HISTORY.md
ng-bubble-cloud Release History
===============================

0.2.0
-----
May 3, 2015
 
Add `tooltip-format-fn` attribute to provide a custom tooltip. (@g-g-n)


0.1.0
-----
April 27, 2014

Initial release.



src/bubble-cloud.js
angular.module('bubbleCloud', [])

.directive('bubbleCloud', function () {

    return {

        restrict: 'E',

        scope: {

            // Either an array of data objects (with groups, labels, and values),
            // or a dictionary of groups of data objects (with labels and values)
            data: '=',

            // Set to 'true' to automatically update when data does
            watch: '@',

            // The attribute containing a data object's value
            valueAttr: '@',

            // The attribute containing a data object's label (optional)
            labelAttr: '@',

            // The attribute containing a data object's label (optional)
            labelAttr: '@',

            // The attribute containing a data object's group
            // (required if data is an array)
            groupAttr: '@',

            // A function which takes a group name and returns the desired
            // fill color. Optional. The default is d3.scale.category20c()
            fillColorFn: '@',

            // A function which takes a group name and returns the desired
            // label color. Optional. The default is black.
            labelColorFn: '@',

            // A function which takes a data object and returns the tooltip
            // text. The default is to combine the label and the value.
            tooltipFormatFn: '@',

            // Overall diameter of the chart, in pixels
            diameter: '@',

            // A function name to publish into the parent scope, to allow
            // reloading the chart (optional)
            renderChartFn: '=?',

        },

        link: function (scope, element, attrs, ctrl) {

            // Set size of element
            element.css('height', scope.diameter + 'px');

            // Publish renderChart into the parent scope
            scope.renderChartFn = ctrl.renderChart;

            // Watch the data, if desired
            if (scope.watch === 'true') {
                scope.$watch('data', ctrl.renderChart, true);
            }

            // Set up the controller
            ctrl.init(element.find('svg'));

            // Render the chart
            ctrl.renderChart();
        },

        template: '<svg id="mychart"></svg>',

        controller: 'chartController',

    };

})

.controller('chartController', function ($scope) {

    // Return a flattened array of objects of this form:
    //
    //   { group: 'Rock', name: 'Rolling Stones, The', value: 12 }
    //
    function get_flattened_data () {

        var data = $scope.data;

        if (_(data).isArray()) {

            var groupAttr = $scope.groupAttr;
            if (! groupAttr) throw new Error('group-attr is required on <bubble>');

            data = _($scope.data).groupBy(function (item) { return item[groupAttr]; });
        }

        var flattened_data = [];

        _(data).each(function (items, group) {

            _(items).each(function (item) {
                flattened_data.push({ group: group, object: item });
            });

        });

        return flattened_data;

    }

    // Initialize the controller with the given SVG element
    // (wrapped in an array)
    this.init = function (svg_element) {

        var valueAttr = $scope.valueAttr;
        if (! valueAttr) throw new Error('value-attr is required on <bubble>');

        var diameter = parseInt($scope.diameter);

        svg_element
            .attr('width', diameter)
            .attr('height', diameter)
            .attr('class', 'bubble');

        $scope.selection = d3.selectAll(svg_element);

        // Create a pack layout
        $scope.pack_layout = d3.layout.pack()
            .sort(null)
            .value(function (datum) {
                return datum.object[valueAttr];
            })
            .size([diameter, diameter])
            .padding(1.5);

        if ($scope.fillColorFn) {
            var fillColorFn = $scope.$parent[$scope.fillColorFn];
            if (! _(fillColorFn).isFunction())
                throw new Error('fill-color-fn attr must be a function in the parent scope');
            $scope.fill_color_fn = fillColorFn;
        } else {
            $scope.fill_color_fn = d3.scale.category20c();
        }

        if ($scope.labelColorFn) {
            var labelColorFn = $scope.$parent[$scope.labelColorFn];
            if (! _(labelColorFn).isFunction())
                throw new Error('label-color-fn attr must be a function in the parent scope');
            $scope.label_color_fn = labelColorFn;
        } else {
            $scope.label_color_fn = function () { return 'black'; };
        }

        if ($scope.tooltipFormatFn) {
            var tooltipFormatFn = $scope.$parent[$scope.tooltipFormatFn];
            if (! _(tooltipFormatFn).isFunction())
                throw new Error('tool-tip-format-fn attr must be a function in the parent scope');
            $scope.tooltip_format_fn = tooltipFormatFn;
        }
    };

    // Get the latest data and render the chart
    this.renderChart = function () {

        // Get the latest data

        var data = { children: get_flattened_data() };

        // Lay out the data

        var node = $scope.selection.selectAll('.node')
            .data($scope.pack_layout.nodes(data).filter(function(d) { return !d.children; }));

        // Handle added nodes

        var enter = node.enter().append('g')
            .attr('class', 'node');
        enter.append('title');
        enter.append('circle');
        enter.append('text')
            .attr('dy', '.3em')
            .style('text-anchor', 'middle')

        // Handle each node

        var valueAttr = $scope.valueAttr;
        var labelAttr = $scope.labelAttr;
        var label_format_fn = $scope.label_format_fn;
        var fill_color_fn = $scope.fill_color_fn;
        var label_color_fn = $scope.label_color_fn;
        var tooltip_format_fn = $scope.tooltip_format_fn;

        node.attr('transform', function (datum) {
            return 'translate(' + datum.x + ',' + datum.y + ')';
        });

        node.select('title')
            .text(function (datum) {
                if (tooltip_format_fn) {
                    return tooltip_format_fn(datum);
                } else {
                    return datum.object[labelAttr] + ': ' + d3.format(',d')(datum.object[valueAttr]);
                }
            });

        node.select('circle')
            .attr('r', function (datum) {
                return datum.r;
            })
            .style('fill', function (datum) {
                return fill_color_fn(datum.group);
            });

        node.select('text')
            .style('fill', function (datum) {
                return label_color_fn(datum.group);
            })
            .text(function (datum) {
                var label = datum.object[labelAttr];
                return label ? label.substring(0, datum.r / 3) : '';
            });

        // Handle removed nodes

        node.exit().remove();
    };

})

;


examples/accounts/accounts.js
angular.module('bubbleCloudTest')

.controller('AccountsCtrl', function ($scope) {
    $scope.assets = [
        { name: 'Checking & Savings', balance: 2000 },
        { name: 'Money Market', balance: 11000 },
        { name: 'Retirement account', balance: 28000 },
        { name: 'Car', balance: 3000 },
    ];
    $scope.liabilities = [
        { name: 'Credit Cards', balance: 5000 },
        { name: 'Laptop', balance: 1500 },
        { name: 'Student Loans', balance: 35000 },
    ];

    $scope.chart_data = {
        assets: $scope.assets,
        liabilities: $scope.liabilities,
    };

    $scope.fill_color = function (group) {
        return group === 'assets' ? '#3eb536' : '#641726';
    }

    $scope.label_color = function (group) {
        return group === 'assets' ? 'black' : '#8c8c8c';
    }

    $scope.tooltip_format = function (datum) {
        return datum.object.name + ': $' + d3.format(',d')(datum.object.balance);
    }

    $scope.new = function (collection) {
        // Pass $scope.assets or $scope.liabilities to create a new item
        collection.push({name: ''});
    }

    $scope.remove = function (collection, item) {
        collection.$remove(item.$id);
    }

})

;
